pkg/upgrade/util.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2022 NVIDIA CORPORATION & AFFILIATES
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kubectl/pkg/drain"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Mellanox/network-operator/pkg/consts"
+)
+
+type StringSet struct {
+	m map[string]bool
+	sync.RWMutex
+}
+
+func NewStringSet() *StringSet {
+	return &StringSet{
+		m: make(map[string]bool),
+	}
+}
+
+// Add item to set
+func (s *StringSet) Add(item string) {
+	s.Lock()
+	defer s.Unlock()
+	s.m[item] = true
+}
+
+// Remove deletes the specified item from the set
+func (s *StringSet) Remove(item string) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.m, item)
+}
+
+// Has looks for item exists in the map
+func (s *StringSet) Has(item string) bool {
+	s.RLock()
+	defer s.RUnlock()
+	_, ok := s.m[item]
+	return ok
+}
+
+// Clear removes all items from the set
+func (s *StringSet) Clear() {
+	s.Lock()
+	defer s.Unlock()
+	s.m = make(map[string]bool)
+}
+
+// KeyedMutex is a struct that provides a per-key synchronized access
+type KeyedMutex struct {
+	mutexes sync.Map // Zero value is empty and ready for use
+}
+
+// Lock locks a mutex, associated with a given key and returns an unlock function
+func (m *KeyedMutex) Lock(key string) func() {
+	value, _ := m.mutexes.LoadOrStore(key, &sync.Mutex{})
+	mtx := value.(*sync.Mutex)
+	mtx.Lock()
+	return func() { mtx.Unlock() }
+}
+
+// NodeUpgradeStateProvider allows for synchronized operations on node objects and ensures that the node,
+// got from the provider, always has the up-to-date upgrade state
+type NodeUpgradeStateProvider struct {
```

Comment:
I don't think this would bring much benefit: we don't need to synchronize access to node operations in other parts of the project and since it's only used for the upgrade flow I think more explicit API is better for code readability